### PR TITLE
Transaction Construction, Pt 7. Fully working transactions

### PR DESCRIPTION
**Pull request by mitschabaude**
This includes the remaining fixes to make transactions work in an end-to-end test, and also to make the old mock API -- `Mina.transaction` -- work again.

This PR also makes improvements to how signing is handled in snarkyjs:
* Instead of passing JSON parties back and forth with OCaml to add signatures, add flexible methods to compute transaction commitments and to sign arbitrary field elements
* Implement "lazy signing", that is, decouple _declaring_ a party that should be signed with actually creating the signature. Important because signatures can only be created once the full transaction is known.

Finally, this also contains a large refactor of `party.ts`, to get rid of unnecessary classes (which are replaced by plain types and functions) and reduce code.

### CODE REVIEW

File: src/lib/party.ts [hunk 1]
```diff
@@ -772,13 +638,31 @@ export class Party {
     return this.body.publicKey;
   }
 
+  sign(privateKey?: PrivateKey) {
```
Comment (bkase) [inline src/lib/party.ts:641]: Can we keep these data types purely functional? This should lead to fewer gotchas for people.

ie. returning a copy of this Party but with the authorization field set

File: src/lib/party.ts [hunk 2]
```diff
@@ -819,27 +703,100 @@ export class Party {
...
+
+type FeePayer = Party & {
+  authorization: Exclude<LazyControl, { kind: 'proof'; value: string }>;
+} & {
+  body: { accountPrecondition: UInt32 };
+};
+
+type Parties = { feePayer: FeePayer; otherParties: Party[] };
+
+function addMissingSignatures(
```
Comment (bkase) [inline src/lib/party.ts:739]: same here -- ideally we have a separate type for a "pending signatures Party" and a "finished Party"

Comment (bkase) [reply src/lib/party.ts:739]: See ExternalTransition.t and ExternalTransition.Validated.t in the OCaml codebase for an example of this

File: src/lib/zkapp.ts [hunk 3]
```diff
@@ -643,38 +661,69 @@ async function callUnproved<S extends typeof SmartContract>(
...
   });
-  let txJson = tx.toJSON();
-  if (zkappKey !== undefined)
-    txJson = Ledger.signOtherParty(txJson, zkappKey, 0);
+  let txJson = tx.sign().toJSON();
   return txJson;
 }
 
-async function signJsonTransaction(
+async function addFeePayer(
```
Comment (bkase) [inline src/lib/zkapp.ts:673]: We should purely functionalize here too -- this one is especially dangerous to use mutation because of the async/await

**Review by bkase — CHANGES_REQUESTED** [review]

File: src/lib/party.ts (hunk 1 above)
Comment (mitschabaude) [reply src/lib/party.ts:641]: Hmm.. So currently, our model for building up the transaction involves actions which mutate the SmartContract's `this.self` party, e.g.
```ts
this.balance.addInPlace(100); // mutates party.body.balance_change
this.x.set(1); // mutates party.body.update.app_state
this.self.update.permissions.setValue(...);
```
I'm not so excited about breaking this flow by introducing a command that replaces the entire party:
```ts
this.self = this.self.sign();
```

So how about this: I could make the `.sign` method on `Party` purely functional, but add another `.sign` method on SmarContract which performs the mutation `this.self = this.self.sign()`?

Comment (mitschabaude) [reply src/lib/party.ts:641]: I started doing this
https://github.com/o1-labs/snarkyjs/pull/123/commits/65937625263c6e9f464cd83f8ebd9aec9d335728
https://github.com/o1-labs/snarkyjs/pull/123/commits/7753a674552cbf9e1585a82eedce620ad3b6d336

I'm still having doubts that this is a good approach

File: src/lib/party.ts (hunk 2 above)
Comment (mitschabaude) [reply src/lib/party.ts:739]: done, I made a type `PartiesValidated` where the signature can't be lazy https://github.com/o1-labs/snarkyjs/pull/123/commits/f10aca039cfda0630ffddbbc9ca3327f814a4d40

File: src/lib/zkapp.ts (hunk 3 above)
Comment (mitschabaude) [reply src/lib/zkapp.ts:673]: done

**Comment by mitschabaude** [comment]
I found a nice way to copy party types: https://github.com/o1-labs/snarkyjs/pull/123/commits/f74daf81673ba83ec38b0c084f2bc44913be6991
With this, I made some steps towards functionalizing: https://github.com/o1-labs/snarkyjs/pull/123/commits/4cc467f388ec2a7687bb06053d76ed777bed2578

I think everything is addressed @bkase

**Review by bkase — APPROVED** [review]

**Pull request merged**